### DISCUSSION TITLE
MutualTLS authentication in C# for httpclient

### DISCUSSION THREAD

**Discussion by vamshimupparaju** [dotnet/runtime#68069] [General] [answered]
Hi Team,

Can you please help me in this from past few days I am trying to resolve the below scenario.

I have private key and certificate key using this I have to do the mutual authentication and send request to one of my end point.

below is the code I am trying to work on but some how I am getting 401 unauthorized issue.

            ****string keyPem = privateKeyString;
            byte[] keyDer = Convert.FromBase64String(keyPem);
            
            X509Certificate2 certWithKey;
            byte[] keycer = Encoding.ASCII.GetBytes(certString);
            X509Certificate2 certOnly = new X509Certificate2(keycer);
            System.Security.Cryptography.RSA rsa = RSA.Create();
         
            rsa.ImportPkcs8PrivateKey(keyDer, out _);
            certWithKey = certOnly.CopyWithPrivateKey(rsa);

            var certificate = new X509Certificate2(certWithKey.Export(X509ContentType.Pfx));

           HttpClientHandler httpClientHandler = new HttpClientHandler();
            httpClientHandler.ClientCertificateOptions = ClientCertificateOption.Manual;
            httpClientHandler.SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls11 | SslProtocols.Tls;

            httpClientHandler.ClientCertificates.Add(certificate);
            httpClientHandler.ClientCertificateOptions = ClientCertificateOption.Manual;

            HttpClient httpClient = new HttpClient(httpClientHandler);

            string _ContentType = "application/x-www-form-urlencoded";
            httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue(_ContentType));
            HttpResponseMessage httpResponseMessage = await httpClient.PostAsync("endpointurl", null);****

Please help me in this whether I am doing anything wrong in the above

@karelz @henning-krause @bartonjs @davidsh @Drawaes @npnelson @stephentoub @jhudsoncedaron

Thanks & Regards,
Vamshi Krishna

**Comment by karelz** [comment, marked as answer]
Duplicate - see https://github.com/dotnet/runtime/issues/68070#issuecomment-1103641069
